streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -764,6 +764,53 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
       } yield pull
     }
 
+  private final def bufferSignal[E1 >: E, A1 >: A](
+    queue: Queue[(Take[E1, A1], Promise[Nothing, Unit])]
+  ): ZManaged[R, Nothing, Pull[R, E1, A1]] =
+    for {
+      as    <- self.process
+      start <- Promise.make[Nothing, Unit].toManaged_
+      _     <- start.succeed(()).toManaged_
+      ref   <- Ref.make(start).toManaged_
+      done  <- Ref.make(false).toManaged_
+      upstream = {
+        def offer(take: Take[E1, A1]): UIO[Unit] = take match {
+          case Take.Value(_) =>
+            for {
+              p     <- Promise.make[Nothing, Unit]
+              added <- queue.offer((take, p))
+              _     <- ref.set(p).when(added)
+            } yield ()
+
+          case _ =>
+            for {
+              latch <- ref.get
```

Comment:
just curious: why is it necessary to wait on error/failure?

Reply:
Continuing my comment from above, we are forcing a complete pause of the pipeline, so that no error is dropped. This way we wait until the whole queue is drained, enqueue the error, and wait until it is dequeued.

Reply:
It is undesirable to silently swallow errors.

Reply:
aaaah yeah I see, in case the queue is dropping when full, you may lose errors indeed. Great 👍